# E-Commerce Marketplace Customer Analytics

## Data Understanding

This notebook focuses on exploring and understanding the raw sales dataset before performing any cleaning, feature engineering, or analysis.

### Objectives
- Load the dataset
- Understand the data structure
- Explore data types
- Assess data quality
- Identify potential issues before cleaning

In [1]:
import pandas as pd
import numpy as np

# Display all columns
pd.set_option("display.max_columns", None)

# Display wider tables
pd.set_option("display.width", None)

## 2. Load Dataset


In [2]:
dataset_path = "../Dataset/Raw/Sales.csv"

df = pd.read_csv(dataset_path, encoding="latin1")

print("Dataset loaded successfully.")

Dataset loaded successfully.


## 3. Dataset Overview

In [3]:
print(f"Number of Rows    : {df.shape[0]}")
print(f"Number of Columns : {df.shape[1]}")

Number of Rows    : 9994
Number of Columns : 21


In [4]:
df.head()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country/Region,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
0,1,INDMKB,8/11/2020,11/11/2020,sas,as,asa,asa,asas,asas,asas,42420.0,INDMKB,FUR-BO-10001798,Furniture,Bookcases,Bush Somerset Collection Bookcase,261.9600,2,0.00,41.9136
1,2,CA-2020-152156,8/11/2020,11/11/2020,Second Class,CG-12520,asasa,Consumer,United States,Henderson,Kentucky,42420.0,South,FUR-CH-10000454,Furniture,Chairs,"Hon Deluxe Fabric Upholstered Stacking Chairs,...",731.9400,3,0.00,219.5820
2,3,CA-2020-138688,12/6/2020,16/6/2020,Second Class,DV-13045,Darrin Van Huff,Corporate,United States,Los Angeles,California,90036.0,West,OFF-LA-10000240,Office Supplies,Labels,Self-Adhesive Address Labels for Typewriters b...,14.6200,2,0.00,6.8714
3,4,US-2019-108966,11/10/2019,18/10/2019,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,FUR-TA-10000577,Furniture,Tables,Bretford CR4500 Series Slim Rectangular Table,957.5775,5,0.45,-383.0310
4,5,US-2019-108966,11/10/2019,18/10/2019,Standard Class,SO-20335,Sean O'Donnell,Consumer,United States,Fort Lauderdale,Florida,33311.0,South,OFF-ST-10000760,Office Supplies,Storage,Eldon Fold 'N Roll Cart System,22.3680,2,0.20,2.5164


In [5]:
df.tail()

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country/Region,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
9989,9990,CA-2018-110422,21/1/2018,23/1/2018,Second Class,TB-21400,Tom Boeckenhauer,Consumer,United States,Miami,Florida,33180.0,South,FUR-FU-10001889,Furniture,Furnishings,Ultra Door Pull Handle,25.248,3,0.2,4.1028
9990,9991,CA-2021-121258,26/2/2021,3/3/2021,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627.0,West,FUR-FU-10000747,Furniture,Furnishings,Tenex B1-RE Series Chair Mats for Low Pile Car...,91.960,2,0.0,15.6332
9991,9992,CA-2021-121258,26/2/2021,3/3/2021,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627.0,West,TEC-PH-10003645,Technology,Phones,Aastra 57i VoIP phone,258.576,2,0.2,19.3932
9992,9993,CA-2021-121258,26/2/2021,3/3/2021,Standard Class,DB-13060,Dave Brooks,Consumer,United States,Costa Mesa,California,92627.0,West,OFF-PA-10004041,Office Supplies,Paper,"It's Hot Message Books with Stickers, 2 3/4"" x 5""",29.600,4,0.0,13.3200
9993,9994,CA-2021-119914,4/5/2021,9/5/2021,Second Class,CC-12220,Chris Cortes,Consumer,United States,Westminster,California,92683.0,West,OFF-AP-10002684,Office Supplies,Appliances,"Acco 7-Outlet Masterpiece Power Center, Wihtou...",243.160,2,0.0,72.9480


In [6]:
df.sample(5, random_state=42)

,Row ID,Order ID,Order Date,Ship Date,Ship Mode,Customer ID,Customer Name,Segment,Country/Region,City,State,Postal Code,Region,Product ID,Category,Sub-Category,Product Name,Sales,Quantity,Discount,Profit
3125,3126,CA-2019-121720,11/6/2019,12/6/2019,First Class,JE-15610,Jim Epp,Corporate,United States,Lakeland,Florida,33801.0,South,OFF-ST-10003816,Office Supplies,Storage,Fellowes High-Stak Drawer Files,563.808,4,0.2,21.1428
1441,1442,CA-2021-128160,19/12/2021,24/12/2021,Second Class,MM-17920,Michael Moore,Consumer,United States,San Francisco,California,94110.0,West,OFF-BI-10001510,Office Supplies,Binders,Deluxe Heavy-Duty Vinyl Round Ring Binder,36.672,2,0.2,11.4600
4510,4511,CA-2020-119935,10/11/2020,14/11/2020,Standard Class,KM-16225,Kalyca Meade,Corporate,United States,Springfield,Missouri,65807.0,Central,FUR-FU-10001085,Furniture,Furnishings,3M Polarizing Light Filter Sleeves,37.300,2,0.0,17.1580
39,40,CA-2019-117415,27/12/2019,31/12/2019,Standard Class,SN-20710,Steve Nguyen,Home Office,United States,Houston,Texas,77041.0,Central,FUR-CH-10004218,Furniture,Chairs,"Global Fabric Manager's Chair, Dark Gray",212.058,3,0.3,-15.1470
4509,4510,CA-2021-167003,23/5/2021,29/5/2021,Standard Class,VS-21820,Vivek Sundaresam,Consumer,United States,Los Angeles,California,90036.0,West,FUR-TA-10001520,Furniture,Tables,"Lesro Sheffield Collection Coffee Table, End T...",171.288,3,0.2,-6.4233


## 4. Data Structure

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9994 entries, 0 to 9993
Data columns (total 21 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Row ID          9994 non-null   int64  
 1   Order ID        9994 non-null   object 
 2   Order Date      9994 non-null   object 
 3   Ship Date       9994 non-null   object 
 4   Ship Mode       9994 non-null   object 
 5   Customer ID     9994 non-null   object 
 6   Customer Name   9994 non-null   object 
 7   Segment         9994 non-null   object 
 8   Country/Region  9994 non-null   object 
 9   City            9994 non-null   object 
 10  State           9994 non-null   object 
 11  Postal Code     9983 non-null   float64
 12  Region          9994 non-null   object 
 13  Product ID      9994 non-null   object 
 14  Category        9994 non-null   object 
 15  Sub-Category    9994 non-null   object 
 16  Product Name    9994 non-null   object 
 17  Sales           9994 non-null   f

In [8]:
df.columns

Index(['Row ID', 'Order ID', 'Order Date', 'Ship Date', 'Ship Mode',
       'Customer ID', 'Customer Name', 'Segment', 'Country/Region', 'City',
       'State', 'Postal Code', 'Region', 'Product ID', 'Category',
       'Sub-Category', 'Product Name', 'Sales', 'Quantity', 'Discount',
       'Profit'],
      dtype='object')

In [9]:
df.dtypes

Row ID              int64
Order ID           object
Order Date         object
Ship Date          object
Ship Mode          object
Customer ID        object
Customer Name      object
Segment            object
Country/Region     object
City               object
State              object
Postal Code       float64
Region             object
Product ID         object
Category           object
Sub-Category       object
Product Name       object
Sales             float64
Quantity            int64
Discount          float64
Profit            float64
dtype: object

## 5. Data Quality Assessment

In [10]:
missing_values = df.isnull().sum()

missing_summary = pd.DataFrame({
    "Missing Values": missing_values,
    "Percentage": (missing_values / len(df)) * 100
})

missing_summary

,Missing Values,Percentage
Row ID,0,0.000000
Order ID,0,0.000000
Order Date,0,0.000000
Ship Date,0,0.000000
Ship Mode,0,0.000000
Customer ID,0,0.000000
Customer Name,0,0.000000
Segment,0,0.000000
Country/Region,0,0.000000
City,0,0.000000


In [11]:
missing_summary[missing_summary["Missing Values"] > 0]

,Missing Values,Percentage
Postal Code,11,0.110066


In [12]:
duplicate_rows = df.duplicated().sum()

print(f"Duplicate Rows : {duplicate_rows}")

Duplicate Rows : 0


In [13]:
df.nunique().sort_values(ascending=False)

Row ID            9994
Profit            7287
Sales             5825
Order ID          5010
Product ID        1862
Product Name      1849
Ship Date         1334
Order Date        1236
Customer Name      798
Customer ID        794
Postal Code        630
City               532
State               50
Sub-Category        17
Quantity            14
Discount            12
Ship Mode            5
Region               5
Segment              4
Category             3
Country/Region       2
dtype: int64

In [14]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
Row ID,9994.0,4997.500000,2885.163629,1.000,2499.25000,4997.5000,7495.750,9994.000
Postal Code,9983.0,55245.233297,32038.715955,1040.000,23223.00000,57103.0000,90008.000,99301.000
Sales,9994.0,229.858001,623.245101,0.444,17.28000,54.4900,209.940,22638.480
Quantity,9994.0,3.789574,2.225110,1.000,2.00000,3.0000,5.000,14.000
Discount,9994.0,0.156203,0.206452,0.000,0.00000,0.2000,0.200,0.800
Profit,9994.0,28.656896,234.260108,-6599.978,1.72875,8.6665,29.364,8399.976


In [15]:
df.describe(include="object").T

,count,unique,top,freq
Order ID,9994,5010,CA-2021-100111,14
Order Date,9994,1236,5/9/2020,38
Ship Date,9994,1334,16/12/2019,35
Ship Mode,9994,5,Standard Class,5968
Customer ID,9994,794,WB-21850,37
Customer Name,9994,798,William Brown,37
Segment,9994,4,Consumer,5190
Country/Region,9994,2,United States,9993
City,9994,532,New York City,915
State,9994,50,California,2001


# Initial Findings & Data Assessment

### Key Findings

- The dataset contains 9,994 sales transaction records and 21 columns.
- No duplicate records were found.
- Only the `Postal Code` column contains missing values (11 records, approximately 0.11%).
- `Order Date` and `Ship Date` are currently stored as text and will be converted to datetime during the data cleaning phase.
- Profit contains both positive and negative values, which is expected in retail sales due to discounts and operational costs.
- The dataset includes 794 unique customers, 1,862 products, and 5,010 unique orders.
- Each order may contain multiple products, which explains why the number of transaction records is greater than the number of unique orders.
- The dataset is suitable for customer analytics, sales analysis, churn prediction, and business intelligence reporting.